File: 01_Statics/00_Solving_A_Rod_IVP/Cpp/main.cpp
#include <iostream>
#include <commonmath.h>
#include <numericalintegration.h>
using namespace ContinuumRobotLibrary;

#ifdef QT_CORE_LIB
#include <simpleplotting.h>
#endif

//Independent Parameters
const double E = 200e9;
const double G = 80e9;
const double rad = 0.001;
const double rho = 8000;
const Vector3d g = 9.81*Vector3d::UnitX();
const double L = 0.5;

//Dependent parameter calculations
const double A = pi*pow(rad,2);
const double I = pi*pow(rad,4)/4;
const double J = 2*I;
const DiagonalMatrix<double, 3> Kse = DiagonalMatrix<double, 3>(G*A,G*A,E*A);
const DiagonalMatrix<double, 3> Kbt = DiagonalMatrix<double, 3>(E*I,E*I,G*J);

//Ordinary differential equation describing elastic rod
VectorXd cosseratRodOde(VectorXd y){
    //Unpack state vector
    Matrix3d R = Map<Matrix3d>(y.segment<9>(3).data());
    Vector3d n = y.segment<3>(12);
    Vector3d m = y.segment<3>(15);

    //Hard-coded material constitutive equation w/ no precurvature
    Vector3d v = Kse.inverse()*R.transpose()*n + Vector3d::UnitZ();
    Vector3d u = Kbt.inverse()*R.transpose()*m;

    //ODEs
    Vector3d p_s = R*v;
    Matrix3d R_s = R*hat(u);
    Vector3d n_s = -rho*A*g;
    Vector3d m_s = -p_s.cross(n);

    //Pack state vector derivative
    VectorXd y_s(18);
    y_s << p_s, Map<VectorXd>(R_s.data(), 9), n_s, m_s;

    return y_s;
}

int main(int, char**){
    //Set initial conditions
    Vector3d p0 = Vector3d::Zero();
    Matrix3d R0 = Matrix3d::Identity();
    Vector3d n0 = Vector3d::UnitY();
    Vector3d m0 = Vector3d::Zero();

    VectorXd y0(18);
    y0 << p0, Map<VectorXd>(R0.data(), 9), n0, m0;

    //Numerically integrate the Cosserat rod equations
    MatrixXd Y = ode4<cosseratRodOde>(y0, L);

    #ifdef QT_CORE_LIB //Plot the solution if Qt is used
    plot(Y.row(1), Y.row(2), "Cosserat Rod IVP Solution", "y (m)", "z (m)");
    #endif

    //Output the data
    std::cout << "Solved rod centerline:\n" << Y.block(0,0,3,Y.cols()) << std::endl;

    return 0;
}

File: 01_Statics/00_Solving_A_Rod_IVP/Python/main.py
import math
import numpy as np
from scipy import integrate
import matplotlib.pyplot as plt


def main():
    # Independent Parameters
    E = 200e9  # Young's modulus
    G = 80e9  # Shear modulus
    r = 0.001  # Cross-sectional radius
    rho = 8000  # Density
    g = np.array([9.81, 0, 0]).T  # Gravitational acceleration
    L = 0.5  # Length(before strain)

    # Dependent Parameters
    A = math.pi * r ** 2  # Cross-sectional area
    I = math.pi * r ** 4 / 4  # Area moment of inertia
    J = 2 * I  # Polar moment of inertia

    Kse = np.diag([G * A, G * A, E * A])  # Stiffness matrices
    Kbt = np.diag([E * I, E * I, G * J])

    # Measured base force and moment
    n0 = np.array([0, 1, 0]).T
    m0 = np.array([0, 0, 0]).T

    # Arbitrary base frame assignment
    p0 = np.array([0, 0, 0]).T
    R0 = np.eye(3)

    # Sub-functions
    def rod_ode(s, y):  # State vector derivative function
        del s  # Integration variable unused in autonomous ODE
        # Unpack state vector
        R = np.reshape(y[3:12], (3, 3))
        n = y[12:15].T
        m = y[15:18].T

        # Constitutive equation
        v = np.linalg.inv(Kse) @ R.T @ n + np.array([0, 0, 1]).T
        u = np.linalg.inv(Kbt) @ R.T @ m

        # Static Cosserat rod equations - system of nonlinear ODEs
        ps = R @ v
        Rs = R @ hat(u)
        ns = -rho * A * g
        ms = -np.cross(ps, n)

        # Pack state vector derivative
        return np.concatenate([ps, np.reshape(Rs, 9), ns, ms]).T

    def hat(y):
        return np.array([[0, -y[2], y[1]],
                         [y[2], 0, -y[0]],
                         [-y[1], y[0], 0]])

    # Numerical Integration
    y0 = np.concatenate([p0.T, np.reshape(R0, 9), n0.T, m0.T]).T  # Combine states into single state vector
    Y = integrate.solve_ivp(rod_ode, (0, L), y0, max_step=0.01).y  # Solve IVP with numerical integration

    # Visualization
    ax = plt.axes(projection='3d')
    ax.plot3D(Y[0, :], Y[1, :], Y[2, :])
    ax.set_xlim([-L / 2, L / 2])
    ax.set_ylim([-L / 2, L / 2])
    ax.set_zlim([0, L])
    plt.title('Rod IVP Solution')
    ax.set_xlabel('x (m)')
    ax.set_ylabel('y (m)')
    ax.set_zlabel('z (m)')
    plt.show()


if __name__ == '__main__':
    main()
